examples/stein_dmm.py

```diff
@@ -0,0 +1,379 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Example: Deep Markov Model inferred using SteinVI
+=================================================
+An implementation of a Deep Markov Model in NumPyro based on reference [1][2] and the Pyro DMM example.
+This is essentially the DKS variant outlined in the paper.
+
+
+**Reference:**
+    1. Pathwise Derivatives for Multivariate Distributions
+    Martin Jankowiak and Theofanis Karaletsos (2019)
+    2. Structured Inference Networks for Nonlinear State Space Models [arXiv:1609.09869]
+    Rahul G. Krishnan, Uri Shalit and David Sontag (2016)
+"""
+import argparse
+
+import jax
+from jax.example_libraries import stax
+import jax.numpy as jnp
+import jax.ops
+
+import numpyro
+from numpyro.contrib.einstein import SteinVI
+from numpyro.contrib.einstein.kernels import RBFKernel
+import numpyro.distributions as dist
+from numpyro.examples.datasets import JSB_CHORALES, load_dataset
+from numpyro.infer import Trace_ELBO
+from numpyro.optim import Adam
+
+
+def _reverse_padded(padded, lengths):
+    def _reverse_single(p, length):
+        new = jnp.zeros_like(p)
+        reverse = jnp.roll(p[::-1], length, axis=0)
+        return jax.ops.index_update(new, jax.ops.index[:], reverse)
+
+    return jax.vmap(_reverse_single)(padded, lengths)
+
+
+def load_data(split="train"):
+    _, fetch = load_dataset(JSB_CHORALES, split=split)
+    lengths, seqs = fetch(0)
+    return (seqs, _reverse_padded(seqs, lengths), lengths)
+
+
+def _one_hot_chorales(seqs, num_nodes=88):
+    return jnp.sum(jnp.array((seqs[..., None] == jnp.arange(num_nodes + 1))), axis=-2)[
+        ..., 1:
+    ]
+
+
+def emitter(hidden_dim1, hidden_dim2, out_dim):
+    """Parameterizes the bernoulli observation likelihood `p(x_t | z_t)`"""
+    return stax.serial(
+        stax.Dense(hidden_dim1),
+        stax.Relu,
+        stax.Dense(hidden_dim2),
+        stax.Relu,
+        stax.Dense(out_dim),
+        stax.Sigmoid,
+    )
+
+
+def transition(gate_hidden_dim, prop_mean_hidden_dim, out_dim):
+    """Parameterizes the gaussian latent transition probability `p(z_t | z_{t-1})`
+    See section 5 in [1].
+
+    **Reference:**
+        1. Structured Inference Networks for Nonlinear State Space Models [arXiv:1609.09869]
+        Rahul G. Krishnan, Uri Shalit and David Sontag (2016)
+    """
+    gate_init_fun, gate_apply_fun = stax.serial(
+        stax.Dense(gate_hidden_dim), stax.Relu, stax.Dense(out_dim), stax.Sigmoid
+    )
+
+    prop_mean_init_fun, prop_mean_apply_fun = stax.serial(
+        stax.Dense(prop_mean_hidden_dim), stax.Relu, stax.Dense(out_dim)
```

Comment:
Not importance: just wonder if it is worth to use either flax or haiku in new tutorials/examples? We used stax in the old days. Maybe we can also use optax for new examples (especially those involve neural networks which is usually used for complicated problems hence needs the flexibility of optax) too. What do you think?

Reply:
It seems that stax has GRUcell that you can use. :D

Reply:
I'll have a looks neat Optax, I haven't worked with it before. Haiku also has a [GRU cell](https://github.com/deepmind/dm-haiku/blob/main/haiku/_src/recurrent.py#L499) C: